lib/winapi/profiling.c

```diff
@@ -3,22 +3,59 @@
 // SPDX-FileCopyrightText: 2019 Stefan Schmidt
 
 #include <profileapi.h>
+#ifdef USE_RDTSC_FOR_FREQ
+#include <synchapi.h>
+#endif
 #include <assert.h>
 #include <stddef.h>
 #include <xboxkrnl/xboxkrnl.h>
 
+#ifdef USE_RDTSC_FOR_FREQ
+static LARGE_INTEGER frequency = {{0, 0}};
+static void __attribute__((constructor)) PrimeQueryPerformanceFrequency ()
+{
+    ULARGE_INTEGER f_rdtsc = {{0, 0}}, s_rdtsc = {{0, 0}};
+    ULONG f_ticks = 0, s_ticks = 0;
+
+    KeEnterCriticalRegion();
+
+    // The values generated after launching aren't accurate, give it time to increment...
+    Sleep(700);
+
+    f_rdtsc.QuadPart = __rdtsc();
+    f_ticks = KeTickCount;
+
+    Sleep(200);
+
+    s_rdtsc.QuadPart = __rdtsc();
+    s_ticks = KeTickCount;
+
+    s_rdtsc.QuadPart -= f_rdtsc.QuadPart;
+    s_rdtsc.QuadPart /= s_ticks - f_ticks;
+
+    frequency.QuadPart = s_rdtsc.QuadPart;
+    frequency.QuadPart *= 1000LL;
```

Comment:
Why multiply here? wouldn't it be more stable to do `s_rdtsc *= 1000LL` above, before the division?

Reply:
It's getting more digits, but not stable, no. Each run is slightly more variable. Wont change.

Reply:
My comment wasn't about clock stability, but https://en.wikipedia.org/wiki/Numerical_stability

You are dividing away up to 3 digits, but at some other point complain about a 1 digit loss in accuracy.

Reply:
> but at some other point complain about a 1 digit loss in accuracy

Let me be clear, as this goes into another comment I made. I don't care about this amount of accuracy. I am only doing it to make *you* happy. If you do not care, then great. Hopefully other maintainers feel the same.

Reply:
> The review mostly focused on getting that time away

Hard disagree. The review has been maintainers attempting to get me to justify every single line of this PR. What percentage would you be happy with? How long are you willing to allow for startup delay? (2nd time I've asked)

Reply:
> you are breaking nxdk for a lot of applications

Is this an assumption or do you have real world examples of this? Have you ran this code?

> timers running a lot less accurate than before

0.0001% less accurate on a stock box. Which I was under the impression you don't care about.
![image](https://github.com/XboxDev/nxdk/assets/858612/51463df4-0bb4-4299-aae0-95eb44f8468f)

Reply:
> Is this an assumption or do you have real world examples of this? Have you ran this code?

I don't need to run this code. I have enough coding experience to know that almost 1 second of additional startup time is bad.

I have a couple of applications in mind which would be broken by this (xbe-loader being one of them).
Generally many network protocols if you tried to forward the network connection between XBEs.
I can also think of things like gamepads. Even without the XID specs I can see that > 500 ms is a huge timeout for a protocol.
Disconnecting a gamepad for almost 1 second likely could cause hardware resets (flickering of LEDs or worse - think of force-feedback).

I'm also familiar enough with these APIs and now how they are being used that the loss in precision matters.

> 0.0001% less accurate on a stock box. Which I was under the impression you don't care about.

Yours is 0.1% as accurate as the existing solution (~1000x worse? Don't quote me on that.. too many numbers being thrown around).

Reply:
> The review has been maintainers attempting to get me to justify every single line of this PR.

Not justify, but modify those lines with errors (or which can be significantly improved with little effort).
That might be close to every line, but that's not an issue with the review - it's an issue with the code.

You should be grateful that people are trying to guide you to better code.

The real issue with this particular review is that valid concerns by two maintainers about the same thing are being ignored (like the numerical stability, which then requires excessive startup delay). Initially we reported some problem at one location, but then the next iteration of the same code had the same issue (the loop was replaced by a single wait which was still too long). And when that wasn't addressed (like the numerical stability), I (we?) sometimes pointed out new problems which arose because the first thing wasn't fixed (such as the 700ms wait with the bad comment).

Pair that with initial skepticism if more than 50% PR is beneficial.
Doing it like the XDK (rdtsc) can be a good idea and it's the initial argument which made this worth reviewing.
Adding the dynamic measurement is likely a bad idea and the way it's done here is poorly executed - especially on closer inspection.

The XDK / rdtsc thing with hardcoded value is mergeable in isolation if the later XDKs also do it (although this might just be a case of function inlining / LTO across the kernel barrier).
It should have been a separate PR and it likely would have been merged immediately.

The dynamic measurement - with some changes - could become acceptable, but it's still worse than the hardcoding and doesn't address the initial use-case you envision (the reasoning / description of which doesn't make sense).. the only use case it does help is with overclocking/underclocking/CPU swaps (which is fine, but not if it has severe drawbacks for everyone else).